Boost: add conditional variants

hainest opened the pull request:
Fixes #1486
Fixes #11467
Closes #31237
Fixes #30663

TODO
- [x] Remove `all_libs`.
    I _think_ this can be done with something like `[v.lib for v in self.variants if v.enabled]` and then call `self.libs()` in `install` (at `for lib in Boost.all_libs:` around line 692).
- [x] Check version ranges for the other variants
- [x] `boost@master` does not build on develop.
    I believe this is due to the `if not with_libs` check in `install`. I have not verified this yet. I will post build error messages later.
- [x] There are many b2 [properties/features](https://www.boost.org/doc/libs/1_79_0/tools/build/doc/html/index.html#bbv2.tutorial.properties) we don't support. See #30496

@trws @alalazo @eugeneswalker @wspear @eschnett @asarkar-parsys

hainest commented:
>There are many b2 [properties/features](https://www.boost.org/doc/libs/1_79_0/tools/build/doc/html/index.html#bbv2.tutorial.properties) we don't support.

boost@master. I have no idea how many of these are useful. TBD.

```
$ find . -iname "*jamfile*" -exec grep -iP "\s*feature\.feature" {} \; -exec echo {} \; -exec echo "" \;
feature.feature phoenix-limit : : free ;
./libs/phoenix/preprocess/Jamfile.v2

feature.feature iterator_debugging
./libs/accumulators/test/Jamfile.v2

feature.feature wavetool 
./libs/wave/tool/build/Jamfile.v2

feature.feature boost.beast.allow-deprecated : on off : propagated composite ;
feature.feature boost.beast.separate-compilation : on off : propagated composite ;
./libs/beast/Jamfile

feature.feature segmented-stacks : on : optional propagated composite ;
feature.feature htm : tsx : optional propagated composite ;
feature.feature valgrind : on : optional propagated composite ;
feature.feature binary-format
feature.feature abi
feature.feature context-impl
./libs/context/build/Jamfile.v2

feature.feature numa : on : optional propagated composite ;
./libs/fiber/build/Jamfile.v2

feature.feature boost.nowide.lfs : auto no : optional propagated ;
./libs/nowide/build/Jamfile.v2

feature.feature nowchar : on :
./libs/lexical_cast/test/Jamfile.v2

feature.feature fusion-limit : : free ;
./libs/fusion/preprocess/Jamfile.v2

feature.feature boost.locale.iconv : on off : optional propagated ;
feature.feature boost.locale.icu : on off :  optional propagated ;
feature.feature boost.locale.posix : on off : optional propagated ;
feature.feature boost.locale.std : on off : optional propagated ;
feature.feature boost.locale.winapi : on off : optional propagated ;
./libs/locale/build/Jamfile.v2

feature.feature <generate-include-all-order> : forward reverse : incidental ;
./libs/units/test_headers/Jamfile.v2

feature.feature <generate-include-all-order> : forward reverse : incidental ;
./libs/spirit/repository/test/test_headers/Jamfile

feature.feature <generate-include-all-order> : forward reverse : incidental ;
./libs/spirit/test/test_headers/Jamfile

    feature.feature known-warnings : hide show : incidental propagated ;
./libs/spirit/Jamfile
```

alalazo commented:
@hainest I added this one to the v0.20.0 which should be completed roughly for ISC23. Let me know if I can help pushing this forward.

hainest commented:
@alalazo All of the code changes are done (I think). I've been trying to build all of the packages that depend on Boost to make sure there are no breakages, but some of them just don't build and I've not had a chance to look at why. I will leave it to you to decide if that's a barrier to merging this.

It's been discussed before, but I want to make sure I enunciate this. By far, the biggest change with this PR is that it's now impossible to request a library/feature that doesn't exist in a particular version of boost.

The current behavior is

```
$ spack spec boost@1.69.0+signals
Input spec
--------------------------------
boost@1.69.0+signals

Concretized
--------------------------------
boost@1.69.0 ... +signals ...
```

but now will fail to concretize because `signals` was removed in 1.68.0. However, it is my contention that the new behavior is correct as the current behavior leads to possible spec reuse when the signals library actually wasn't installed.

hainest commented:
@spackbot re-run pipeline

hainest committed:
Move patches into separate module, subdirectory

This makes the package file easier to read and cleans up the package
directory.

* Put remote patches into local files

    This allows building on air-gapped networks.

* Use more uniform names for patch files

* Extend phoenix patch until 1.83.0

    It wasn't completely fixed until that release.

* Fill in missing patch SHAs

* Remove references to svn issues

    The svn server has been taken offline, so these URLs are no longer
    valid.

hainest committed:
Extract visibility variant

This also enforces the version constraint.

hainest committed:
Extract context, context-impl variants

hainest committed:
Extract container variant

hainest committed:
Extract coroutine variant

hainest commented:
1.86.0 added

1. CRC requires c++ >= 11
2. Format requires c++ >= 11
3. UUID requires c++ >= 11
4. Graph requires c++ >= 14
5. JSON deprecated gcc <= 5

hainest committed:
Extract clanglibcpp variant, add conflict for gcc

gcc doesn't support libc++.

hainest committed:
Extract tagged{versioned}layout variants

hainest committed:
Extract debug variant

hainest committed:
Extract threading variants

hainest committed:
Extract pic variant, enable by default

It makes no sense to not have it on by default since +shared is also
enabled by default. I would remove it as a variant, but there might be
someone who is explicity setting it.

hainest committed:
Extract, cleanup numpy variant

hainest committed:
Extract, cleanup python variant

hainest committed:
Extract, cleanup mpi variant

This slightly changes the logic of the conflict with 1.64.0 when
+python is enabled, but is now consistent with the comment. That is,
the conflict happens with @1.64.0+python+mpi instead of just
@1.64.0+python.

hainest committed:
Extract, cleanup iostreams variant

hainest committed:
Extract, cleanup fiber variant

hainest committed:
Extract, modify signals{2} variants

This changes the logic on Windows such that the correct signals library
is always required. This is because `all_libs` will go away and
maintaining a list of Windows libraries that actually need signals{2}
is a substantial maintenance burden.

hainest committed:
Extract atomic variant

hainest committed:
Extract charconv variant

hainest committed:
Extract chrono variant

hainest committed:
Extract cobalt variant, add leaf variant

hainest committed:
Extract contract variant

hainest committed:
Extract date_time variant

hainest committed:
Extract exception variant

hainest committed:
Extract filesystem variant

hainest committed:
Extract graph variant

hainest committed:
Extract graph_parallel variant

hainest committed:
Extract locale variant, don't install by default

It turns out to be very expensive to build icu4c and its dependencies.

hainest committed:
Extract log variant

hainest committed:
Extract math variant

hainest committed:
Extract nowide variant

hainest committed:
Extract program_options variant

hainest committed:
Extract random variant

hainest committed:
Extract regex variant

hainest committed:
Extract serialization variant

hainest committed:
Extract stacktrace variant

hainest committed:
Extract thread variant

hainest committed:
Extract timer variant

hainest committed:
Extract type_erasure variant

hainest committed:
Extract url variant

hainest committed:
Extract wave variant

hainest committed:
Remove old filtering, use new variants for library building options

* Remove 'with_libs' filtering
    The conditional variants handle this.

* Use set of buildable libraries from conditional variants in 'with_libs'

* Move 'with_libs' calculation to 'determine_booststrap_options'
    It's only used in that one place.

* Replace use of 'all_libs' with 'variants.libraries_to_build'

    This ensures that all libraries are accounted for and only the
    ones that are actually needed for compilation are used for
    those checks.

hainest committed:
Don't double-quote cxxflags passed to b2

b2 already quotes them, so passing multiple values causes them to be
passed to the compiler as a quoted string like '-fFoo -fBar'.

hainest committed:
Extract test variant

hainest committed:
Extract system variant

hainest committed:
Add updates from 1.75.0 release notes

hainest committed:
Add updates from 1.78.0 release notes

hainest committed:
Add updates from 1.81.0 release notes

hainest committed:
Add updates from 1.82.0 release notes

hainest committed:
Add updates from 1.83.0 release notes

hainest committed:
Add updates from 1.85.0 release notes

bernhardkaindl reviewed (commented):
@hainest: You need to resolve this conflict with develop to continue this PR!

hainest committed:
sarus: update Boost dependency

yap and hana are enabled by default in 1.65.0, but require c++14.

hainest committed:
Add updates from 1.80.0 release notes

hainest committed:
Add updates from 1.84.0 release notes

Many libraries now require c++11, so it's easier to just make the
entire package require c++11.

hainest committed:
Add README

hainest committed:
Add boost-specific variant type

Variants are sticky by default to prevent the solver from changing them.
In particular, it prevents the solver from changing cxxstd when
a defaulted variant requires a newer version than the user specified.

hainest committed:
Extract cxxstd variant, set default to 14

Restricting to c++11 sets the default version to 1.60.0 because most
libraries after that require c++14. The default of c++14 makes 1.83.0
the default version.

hainest committed:
Extract shared variant, require when building shared libraries

+shared~pic makes no sense.

hainest committed:
Extract, cleanup icu variant

hainest committed:
Add all variants up to 1.75.0

These libraries are intentionally left out because they don't have any
constraints as of 1.87.0:

accumulators, array, assert, assign, bimap, call_traits,
circular_buffer, compatibility, compressed_pair, concept_check, config,
dynamic_bitset, enable_if, flyweight, foreach, function_types,
fusion, in_place_factory, interprocess, interval, intrusive, io,
iterator, lambda, min_max, multi_array, multi_index, numeric_conversion,
optional, pool, proto, ptr_container, range, rational, ref, result_of,
scope_exit, smart_ptr, statechart, static_assert, string_algo, swap,
tokenizer, tribool, tuple, type_traits, ublas, units, utility,
value_initialized, variant, xpressive

hainest committed:
Add updates from 1.76.0 release notes

hainest committed:
Add updates from 1.77.0 release notes

hainest committed:
Add updates from 1.79.0 release notes

hainest committed:
Add updates from 1.87.0 release notes

hainest committed:
Add updates from 1.86.0 release notes

hainest committed:
Remove with_default_variants for packages with implicit dependence

hainest committed:
Remove with_default_variants

hainest committed:
Extract JSON variant

hainest committed:
Use 'spec.satisfies' instead of 'in spec'

This was introduced by 530639e15fb7.

hainest committed:
Extract patch() into patches.py

hainest committed:
cleanup conflicts

hainest committed:
Extract toolset

hainest committed:
Refactor bootstrapping

hainest committed:
Use boost.io for source downloads

hainest committed:
Move add_buildopt_symlinks out of the way

hainest committed:
Refactor bjam

hainest committed:
Make MPI install process more explicit

This technically adds an additional check for MPI, but it makes it
clearer as to why this is needed. In the even that both threading
types are wanted, it will invoke `b2 --install` only once.

hainest committed:
Improve constraints on patches

There's no need to apply patches that don't pertain to the current set
of variants.

hainest committed:
Add tests

hainest commented:
@alalazo @tldahlgren

I can't figure out why the macos [check](https://github.com/spack/spack/actions/runs/13576847873/job/37954971365?pr=30627) is failing. `etc/spack/defaults/darwin/packages.yaml` doesn't contain any mention of boost, and the new boost/package.py doesn't mention signals at all. Any thoughts would be greatly appreciated.

I would like to add the tests from "boost/tests/" to the CI pipeline to be triggered when this package is modified. I've read through the [CI](https://spack.readthedocs.io/en/latest/pipelines.html) docs, but I haven't been able to determine how to wire into the CI run from this repo.

 I am doing regression tests for all packages that depend on boost. Getting those all green will be a huge step forward.

Finally, I need to add a README to help users with this package. There is a fairly complex interplay between the philosophy of variants in this package compared to the philosophy of b2. It's unavoidable given the way the two installers work. I'll also add lots of details on how to add new versions, variants, patches, and tests.

Thanks so much for your help with all of this!

msimberg commented on var/spack/repos/builtin/packages/boost/boostorg/variants.py:
```diff
@@ -0,0 +1,1479 @@
+        ],
+        description="A parser combinator library for building complex parsers",
+    )
+
```
```suggestion
    variants.add(
        "mqtt5",
        when="@1.88.0:",
        buildable="@1.88.0:",
        description="A C++17 MQTT client",
    )
```
and possibly `hash2`? It looks like you're modeling all the libraries now, right? Not only the ones that require building with `--with-libraries`?

From Boost 1.88.0: https://github.com/spack/spack/pull/50158.

hainest replied:
> and possibly `hash2`?

Yup. I just haven't added 1.88 yet.

>It looks like you're modeling all the libraries now, right? Not only the ones that require building with `--with-libraries`?

I model any library that has a constraint. Which subset that is changes by version. For example, in 1.88.0 the Iterator library dropped support for C++03.

hainest committed:
Only patch numpy when the variant is active

hainest committed:
Remove with_default_variants for packages with explicit dependence
